zerver/lib/actions.py

```diff
@@ -3562,6 +3563,20 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
             subs_to_deactivate.append(sub_info)
             sub_ids_to_deactivate.append(sub_info.sub.id)
 
+    stream_ids = [stream_id for stream_id in stream_dict.keys()]
+    stream_admin_dict = get_stream_admin_dict_for_streams(stream_ids)
+    stream_admin_count_dict: Dict[int, int] = {}
+    for stream_id in stream_admin_dict.keys():
+        stream_admin_count_dict[stream_id] = len(stream_admin_dict[stream_id])
+
+    for sub in subs_to_deactivate:
+        user_profile_id = sub.user.id
+        stream_id = sub.stream.id
+        if user_profile_id in stream_admin_dict[stream_id]:
+            if stream_admin_count_dict[stream_id] == 1:
+                raise JsonableError(_("Cannot remove all stream administrators."))
+            stream_admin_count_dict[stream_id] -= 1
```

Comment:
Let's move this block to a function `check_last_stream_admin_cannot_unsubscribe`.  I think we could also implement that function a bit more simply as e.g.:
```
stream_admin_dict = get_stream_admin_dict_for_streams(stream_ids) # Returns a Dict[id, Set[id]]
for sub in subs_to_deactivate:
    stream_admin_dict[stream_id].remove(sub.user_profile_id)

for (stream_id, remaining_admins) in stream_admin_dict.items():
    if len(remaining_admins) == 0:
        raise ...
```